src/mapeo-project.js

```diff
@@ -548,6 +565,59 @@ export class MapeoProject extends TypedEmitter {
   get $icons() {
     return this.#iconApi
   }
+
+  async $leave() {
+    await assertDeviceCanLeaveProject(this.#deviceId, this.#capabilities)
+
+    // update client database (delete all encryption keys except auth)
+    const encoded = ProjectKeys.encode({
+      projectKey: Buffer.from(this.#projectId, 'hex'),
+      projectSecretKey: this.#projectSecretKey,
+      encryptionKeys: { auth: this.#encryptionKeys.auth },
+    }).finish()
+
+    const nonce = projectIdToNonce(this.#projectId)
+
+    // 1. Update relevant shared database tables
+    this.#sharedDb
+      .update(projectKeysTable)
+      .set({
+        keysCipher: this.#keyManager.encryptLocalMessage(
+          Buffer.from(encoded.buffer, encoded.byteOffset, encoded.byteLength),
+          nonce
+        ),
+      })
+      .where(eq(projectKeysTable.projectId, this.#projectId))
+      .run()
+
+    this.#sharedDb
+      .delete(projectSettingsTable)
+      .where(eq(projectSettingsTable.docId, this.#projectId))
+      .run()
+
+    // 2. Clear data from cores
+    // TODO: only clear synced data
+    const namespacesWithoutAuth =
+      /** @satisfies {Exclude<import('./core-manager/index.js').Namespace, 'auth'>[]} */ ([
+        'config',
+        'data',
+        'blob',
+        'blobIndex',
+      ])
+
+    await Promise.all(
+      namespacesWithoutAuth.map((namespace) =>
+        this.#coreManager.deleteData(namespace, { deleteOwn: true })
+      )
+    )
+
+    // 3. Clear data from project database
```

Comment:
Haven't implemented yet because I have questions about how this should be done. Is it just a naive db.delete(_table_)? (where `db = new Database(this.#sqlite)`)

Wondering if the deletion of data in sqlite should be handled by something like `CoreManager.deleteData()`? A little bit tangled with the different boundaries and what should be responsible for this 😅